src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ExhaustivenessComputer.java

```diff
@@ -673,5 +782,432 @@ public String toString() {
+                Type t1 = bp1.type();
+                Type t2 = bp2.type();
+                boolean t1IsImportantRecord =
+                        (t1.tsym.flags_field & RECORD) != 0 &&
+                        hasMatchingRecordPattern(basePatterns, missingPatterns, bp1);
+                boolean t2IsImportantRecord =
+                        (t2.tsym.flags_field & RECORD) != 0 &&
+                        hasMatchingRecordPattern(basePatterns, missingPatterns, bp2);
+                if (t1IsImportantRecord && !t2IsImportantRecord) {
+                    return false;
+                }
+                if (!t1IsImportantRecord && t2IsImportantRecord) {
+                    return true;
+                }
+                if (!types.isSameType(t1, t2) && types.isSubtype(t1, t2)) {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+    /*
+     * Do the {@code basePatterns} have a record pattern at a place that corresponds to
+     * position of pattern {@code query} inside {@code missingPatterns}?
+     */
+    private boolean hasMatchingRecordPattern(Set<? extends PatternDescription> basePatterns,
+                                             Set<? extends PatternDescription> missingPatterns,
+                                             PatternDescription query) {
+        PatternDescription root = findRootContaining(missingPatterns, query);
+
+        if (root == null) {
+            return false;
+        }
+        return basePatternsHaveRecordPatternOnThisSpot(basePatterns, root, query);
+    }
+    //where:
+        private PatternDescription findRootContaining(Set<? extends PatternDescription> rootPatterns,
+                                                      PatternDescription added) {
+            for (PatternDescription pd : rootPatterns) {
+                if (isUnderRoot(pd, added)) {
+                    return pd;
+                }
+            }
+
+            //assert?
+            return null;
+        }
+
+        private boolean basePatternsHaveRecordPatternOnThisSpot(Set<? extends PatternDescription> basePatterns,
+                                                                PatternDescription rootPattern,
+                                                                PatternDescription added) {
+            if (rootPattern == added) {
+                return basePatterns.stream().anyMatch(pd -> pd instanceof RecordPattern);
+            }
+            if (!(rootPattern instanceof RecordPattern rootPatternRecord)) {
+                return false;
+            }
+            int index = -1;
+            for (int c = 0; c < rootPatternRecord.nested.length; c++) {
+                if (isUnderRoot(rootPatternRecord.nested[c], added)) {
+                    index = c;
+                    break;
+                }
+            }
+            Assert.check(index != (-1));
```

Comment:
Small suggestion for understanding better what the assertion would imply.

```suggestion
            // 'index' must be one of rootPatternRecord.nested; if not, `isUnderRoot` is inconsistent.
            Assert.check(index != (-1)); 
```